hail/src/main/scala/is/hail/expr/ir/IR.scala

```diff
@@ -330,12 +330,12 @@ final case class Die(message: IR, _typ: Type) extends IR
 final case class ApplyIR(function: String, args: Seq[IR]) extends IR {
   var conversion: Seq[IR] => IR = _
 
-  lazy val explicitNode: IR = {
-    val refs = args.map(a => Ref(genUID(), a.typ)).toArray
-    var body = conversion(refs).deepCopy()
+  private lazy val refs = args.map(a => Ref(genUID(), a.typ)).toArray
+  lazy val body: IR = conversion(refs).deepCopy()
 
+  lazy val explicitNode: IR = {
     // foldRight because arg1 should be at the top so it is evaluated first
-    refs.zip(args).foldRight(body) { case ((ref, arg), bodyIR) => Let(ref.name, arg, bodyIR) }
+    Optimize(refs.zip(args).foldRight(body) { case ((ref, arg), bodyIR) => Let(ref.name, arg, bodyIR) })
```

Comment:
This will return to do additional optimization passes in Emit right? It feels most intuitive to me that once you're emitting code optimization is already done.

Reply:
this is looked at in simplify -- https://github.com/hail-is/hail/blob/f160d4a87d6db708dbcaa3c28dc7ad8be0295e7e/hail/src/main/scala/is/hail/expr/ir/Simplify.scala#L185

Reply:
It's definitely not ideal to optimize here, but were emitting unoptimized code.

Reply:
@cseed second opinion welcome